### PR TITLE
Disable Hyperthreading

### PR DESCRIPTION
**Pull request by sean-smith**
In your config file, set `disable_hyperthreading = true` (default is false).

And set the `compute_instance_type` and `master_instance_type` to an instance type that supports CpuOptions.

```ini
[cluster test]
...
master_instance_type = c5.xlarge
compute_instance_type = c5.2xlarge
disable_hyperthreading = true
```

Once you ssh onto the cluster, you'll see each instance has vCpu / 2 number of cores:

```bash
$ lscpu
...
Thread(s) per core:    1
Core(s) per socket:    2
$ qhost
HOSTNAME                ARCH         NCPU NSOC NCOR NTHR  LOAD  MEMTOT  MEMUSE  SWAPTO  SWAPUS
----------------------------------------------------------------------------------------------
global                  -               -    -    -    -     -       -       -       -       -
ip-172-31-3-84          lx-amd64        4    1    4    4  0.01   14.9G  330.8M     0.0     0.0
```

Signed-off-by: Sean Smith <seaam@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/pcluster/config/mappings.py [hunk 1]
```diff
@@ -477,6 +478,12 @@
         "s3_read_write_resource": {
             "cfn_param_mapping": "S3ReadWriteResource",  # TODO add validator
         },
+        "disable_hyperthreading": {
+            "type": DisableHyperThreading,
```
Comment (enrico-usai) [inline cli/pcluster/config/mappings.py]: Should be: `DisableHyperThreadingParam` to be consistent with the others.

File: cli/pcluster/config/mappings.py [hunk 2]
```diff
@@ -477,6 +478,12 @@
...
+            "default": "false",
+            "allowed_values": ["true", "false"],
```
Comment (enrico-usai) [inline cli/pcluster/config/mappings.py]: The new parameter is a boolean so you should define the new param type by inheriting the `BoolParam` type :
` class DisableHyperThreading(BoolParam):  `

Then you don't need to specify the `allowed_values` and the `default` can be set to `False`.

```python
 "disable_hyperthreading": {
     "type": DisableHyperThreadingParam,
     "default": False,
     "cfn_param_mapping": "Cores",
}
```

File: cli/pcluster/config/param_types.py [hunk 3]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+    Class to manage the disable_hyperthreading configuration parameter.
+
+    We need this class in order to convert the boolean disable_hyperthreading = [true/false] into Cores.
+    """
+
+    def from_cfn_params(self, cfn_params):
+        """Initialize param value by parsing the right CFN input."""
+        cfn_converter = self.definition.get("cfn_param_mapping", None)
+        if cfn_converter and cfn_params:
+            cores = int(get_cfn_param(cfn_params, cfn_converter).split(",")[0])
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: You need a test to verify the value returned from `get_cfn_param` because if the param is not present or NONE (e.g. an old cluster without the CFN parameter) we can have an `IndexError`.

File: cli/pcluster/config/param_types.py [hunk 4]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+            self.value = "false"
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: When you inherit from BoolParam you can use `False` and `True`.

Then you can compact the code, by writing:
`self.value = cores > 0`

File: cli/pcluster/config/param_types.py [hunk 5]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+        """
+        Return a tuple (cores_master,cores_compute) if disable_hyperthreading = true.
+
+        :return: string (cores_master,cores_compute)
+        """
+
+        cfn_params = {}
+
+        cluster_config = self.pcluster_config.get_section(self.section_key)
+        if cluster_config.get_param_value("disable_hyperthreading") == "true":
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: BoolParam --> `if cluster_config.get_param_value("disable_hyperthreading"):`

File: cli/pcluster/config/param_types.py [hunk 6]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+            master_instance_type = cluster_config.get_param_value("master_instance_type")
+            compute_instance_type = cluster_config.get_param_value("compute_instance_type")
+            master_cores = get_instance_vcpus(self.pcluster_config.region, master_instance_type) // 2
+            compute_cores = get_instance_vcpus(self.pcluster_config.region, compute_instance_type) // 2
+            if master_cores < 0 or compute_cores < 0:
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: Minor: I'd add a newline between variables initialization and their usage but it's up to you.

File: cli/pcluster/config/param_types.py [hunk 7]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+                error(
+                    "Unable to get number of vcpus for {0} instance. ".format(
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: I'd add an explanation about the configuration parameter causing the issue. 
In this way they could try to change the parameter and go ahead while waiting for an eventual fix.

Then what do you think about add the format at the end of the whole error message string and not in the middle?

File: cli/pcluster/config/param_types.py [hunk 8]
```diff
@@ -675,6 +683,59 @@ def to_file(self, config_parser):
...
+                        master_instance_type if master_cores < 0 else compute_instance_type
+                    )
+                    + "Please open an issue https://github.com/aws/aws-parallelcluster/issues"
+                )
+            cfn_params.update({self.definition.get("cfn_param_mapping"): "{0},{1}".format(master_cores, compute_cores)})
+
+            # Check to see if cfn_scheduler_slots is set
+            extra_json = cluster_config.get_param_value("extra_json")
+            if extra_json:
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: This should be moved in the validator,. In this way the user will be notified at validation time and not at conversion time.

File: cli/pcluster/config/pcluster_config.py [hunk 9]
```diff
@@ -344,18 +345,38 @@ def __check_account_capacity(self):  # noqa: C901
...
+                compute_subnet = master_subnet
 
             placement_group = cluster_section.get_param_value("placement_group")
+            disable_hyperthreading = cluster_section.get_param_value("disable_hyperthreading")
+            master_vcpus = get_instance_vcpus(self.region, master_instance_type)
+            compute_vcpus = get_instance_vcpus(self.region, compute_instance_type)
 
             boto3.client("ec2").run_instances(
-                InstanceType=instance_type,
+                InstanceType=master_instance_type,
```
Comment (enrico-usai) [inline cli/pcluster/config/pcluster_config.py]: Could you add two comments before the run_instances describing why we are doing two tests (I know it is to test master and compute but we have to check the code to find the differences).

Then we can define the variables for `CpuOptions` and `Placement` in a single place before the two run_instances.

File: cli/pcluster/config/pcluster_config.py [hunk 10]
```diff
@@ -364,6 +385,8 @@ def __check_account_capacity(self):  # noqa: C901
             message = e.response.get("Error").get("Message")
             if code == "DryRunOperation":
                 pass
+            elif code == "UnsupportedOperation":
+                error(message.replace("CpuOptions", "disable_hyperthreading"))
```
Comment (enrico-usai) [inline cli/pcluster/config/pcluster_config.py]: I think we should write our custom message instead of replacing `CpuOptions` from a message that can change in future.

File: cli/tests/pcluster/config/defaults.py [hunk 11]
```diff
@@ -122,6 +122,7 @@
     "efs_settings": None,
     "raid_settings": None,
     "fsx_settings": None,
+    "disable_hyperthreading": "false",
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/defaults.py]: Please move it before the `*_settings` parameters. Then let's use boolean.

File: cli/tests/pcluster/config/test_section_cluster.py [hunk 12]
```diff
@@ -253,6 +254,7 @@ def test_cluster_section_from_240_cfn(mocker, cfn_params_dict, expected_section_
                 "VolumeIOPS": "NONE,NONE,NONE,NONE,NONE",
                 "VolumeSize": "20,NONE,NONE,NONE,NONE",
                 "VolumeType": "gp2,NONE,NONE,NONE,NONE",
+                "Cores": "-1,-1",
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/test_section_cluster.py]: You should remove the "Cores" parameter from this test, because it is the `test_cluster_section_from_231_cfn` because it is testing a set of input parameters coming from a cluster created with the 2.3.1 version

File: cli/tests/pcluster/config/test_section_cluster.py [hunk 13]
```diff
@@ -337,6 +339,7 @@ def test_cluster_section_from_231_cfn(mocker, cfn_params_dict, expected_section_
                 "VolumeIOPS": "100, 100, 100, 100, 100",
                 "VolumeSize": "20, 20, 20, 20, 20",
                 "VolumeType": "gp2, gp2, gp2, gp2, gp2",
+                "Cores": "-1,-1",
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/test_section_cluster.py]: Same as before.

File: cli/tests/pcluster/config/test_section_cluster.py [hunk 14]
```diff
@@ -421,6 +424,7 @@ def test_cluster_section_from_210_cfn(mocker, cfn_params_dict, expected_section_
...
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/test_section_cluster.py]: Same

File: cli/tests/pcluster/config/test_section_cluster.py [hunk 15]
```diff
@@ -167,6 +167,7 @@ def test_cluster_section_from_241_cfn(mocker, cfn_params_dict, expected_section_
...
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/test_section_cluster.py]: We should remove the "Cores" parameter from this test, it is the `test_cluster_section_from_241_cfn` because it is testing a set of input parameters coming from a cluster created with the 2.4.1 version.

Please add a new test on top for the 2.5.0 version with the new parameters (e.g. the IAMPolicies too)

File: cloudformation/aws-parallelcluster.cfn.json [hunk 16]
```diff
@@ -320,6 +320,11 @@
       "Type": "String",
       "Default": "NONE,NONE,NONE,NONE,NONE,NONE,NONE,NONE"
     },
+    "Cores": {
+      "Description": "Comma seperated string of [master cores],[compute cores]. If set to -1, no CpuOptions are set.",
+      "Type": "String",
```
Comment (enrico-usai) [inline cloudformation/aws-parallelcluster.cfn.json]: Could you use `Type": "CommaDelimitedList",` ?

File: cli/pcluster/config/param_types.py (hunk 8 above)
Comment (enrico-usai) [reply cli/pcluster/config/param_types.py]: Please add the validator and also the unit test for it.

Comment (sean-smith) [reply cli/pcluster/config/param_types.py]: moved to validator

TODO: add unit tests

File: cli/pcluster/config/pcluster_config.py (hunk 10 above)
Comment (enrico-usai) [reply cli/pcluster/config/pcluster_config.py]: Ok, thanks for the detailed explanation! 👍  I agree we can keep the `message.replace()` for now.

File: cli/pcluster/config/param_types.py [hunk 17]
```diff
@@ -673,6 +681,55 @@ def to_file(self, config_parser):
...
+    We need this class in order to convert the boolean disable_hyperthreading = [true/false] into Cores.
+    """
+
+    def from_cfn_params(self, cfn_params):
+        """Initialize param value by parsing the right CFN input."""
+        cfn_converter = self.definition.get("cfn_param_mapping", None)
+        if cfn_converter and cfn_params:
+            cores = get_cfn_param(cfn_params, cfn_converter)
+            if cores and cores != "NONE":
+                cores = int(cores.split(",")[0])
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: I'd add a try catch block for ValueError and IndexError here

File: cli/pcluster/config/param_types.py [hunk 18]
```diff
@@ -673,6 +681,55 @@ def to_file(self, config_parser):
...
+                error(
+                    "For disable_hyperthreading, unable to get number of vcpus for {0} instance. Please open an issue {1}".format(
+                        master_instance_type if master_cores < 0 else compute_instance_type,
+                        "https://github.com/aws/aws-parallelcluster/issues",
+                    )
+                )
+            cfn_params.update({self.definition.get("cfn_param_mapping"): "{0},{1}".format(master_cores, compute_cores)})
+
+        else:
+            cfn_params.update({self.definition.get("cfn_param_mapping"): "-1,-1"})
```
Comment (enrico-usai) [inline cli/pcluster/config/param_types.py]: Minor: You could use -1,-1 as default values in the cfn_params initialization and get ride of the else block. (up to you)

File: cli/pcluster/config/validators.py [hunk 19]
```diff
@@ -192,6 +192,21 @@ def fsx_storage_capacity_validator(param_key, param_value, pcluster_config):
...
 
 
+def disable_hyperthreading_validator(param_key, param_value, pcluster_config):
+    errors = []
+    warnings = []
+
+    if param_value:
+        # Check to see if cfn_scheduler_slots is set
+        extra_json = pcluster_config.get_param_value("extra_json")
+        if extra_json:
```
Comment (enrico-usai) [inline cli/pcluster/config/validators.py]: you could compact this with:
```py
if extra_json and extra_json.get("cluster") and extra_json.get("cluster").get("cfn_scheduler_slots"):
     errors.append("cfn_scheduler_slots cannot be set in addition to disable_hyperthreading = true")
```

File: cli/tests/pcluster/config/test_section_cluster.py (hunk 15 above)
Comment (enrico-usai) [reply cli/tests/pcluster/config/test_section_cluster.py]: You missed it ;)

File: cloudformation/aws-parallelcluster.cfn.json [hunk 20]
```diff
@@ -320,6 +320,11 @@
       "Type": "String",
       "Default": "NONE,NONE,NONE,NONE,NONE,NONE,NONE,NONE"
     },
+    "Cores": {
+      "Description": "Comma seperated string of [master cores],[compute cores]. If set to -1, no CpuOptions are set.",
```
Comment (enrico-usai) [inline cloudformation/aws-parallelcluster.cfn.json]: typo: separated
typo: If set to -1,-1 ?

File: cli/tests/pcluster/config/test_section_cluster/test_cluster_from_file_to_cfn/pcluster.config.ini [hunk 21]
```diff
@@ -211,4 +212,5 @@ extra_json = { "cluster" : { "cfn_scheduler_slots" : "cores" } }
 additional_cfn_template = https://test
 tags = {"test": "test"}
 custom_chef_cookbook = https://test
-custom_awsbatch_template_url = https://test
\ No newline at end of file
+custom_awsbatch_template_url = https://test
+disable_hyperthreading = false
```
Comment (enrico-usai) [inline cli/tests/pcluster/config/test_section_cluster/test_cluster_from_file_to_cfn/pcluster.config.ini]: Could you add a new test with `disable_hyperhreading = true` or change this one?

File: cli/pcluster/config/mappings.py [hunk 22]
```diff
@@ -477,6 +478,11 @@
         "s3_read_write_resource": {
             "cfn_param_mapping": "S3ReadWriteResource",  # TODO add validator
         },
+        "disable_hyperthreading": {
```
Comment (enrico-usai) [inline cli/pcluster/config/mappings.py]: Could you please add unit tests for this new parameter in the `test_section_cluster.py --> test_cluster_param_from_file`  ?

Comment (enrico-usai) [reply cli/pcluster/config/mappings.py]: Probably you could also add muliple tests in the `test_section_cluster.py --> test_cluster_section_to_file` since the disable_hyperthreading is strictly related to the extra_json and we can test the different combinations.

File: cli/pcluster/config/param_types.py (hunk 8 above)
Comment (sean-smith) [reply cli/pcluster/config/param_types.py]: unit tests added

File: cli/pcluster/config/mappings.py (hunk 22 above)
Comment (sean-smith) [reply cli/pcluster/config/mappings.py]: I added tests to `test_cluster_param_from_file()` and `test_cluster_section_to_file()`

File: cli/pcluster/config/mappings.py [hunk 23]
```diff
@@ -462,24 +464,30 @@
...
-                "cfn_param_mapping": "EC2IAMPolicies",
-                "validators": [ec2_iam_policies_validator],
-            }),
-            ("s3_read_resource", {
-                "cfn_param_mapping": "S3ReadResource",  # TODO add validator
-            }),
-            ("s3_read_write_resource", {
-                "cfn_param_mapping": "S3ReadWriteResource",  # TODO add validator
-            }),
+            ("proxy_server", {"cfn_param_mapping": "ProxyServer"}),
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/pcluster/config/mappings.py]: Changed back

File: cli/pcluster/config/pcluster_config.py [hunk 24]
```diff
@@ -329,40 +329,83 @@ def __check_account_capacity(self):  # noqa: C901
...
+            MinCount=max_size,
+            MaxCount=max_size,
+            ImageId=get_latest_alinux_ami_id(),
+            SubnetId=compute_subnet,
+            CpuOptions=compute_cpu_options,
+            Placement=compute_placement_group,
+            DryRun=True,
+        )
+
+    def __ec2_run_instance(self, max_size, **kwargs):
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/pcluster/config/pcluster_config.py]: It's used in the error messages, line 417

File: cli/tests/pcluster/config/test_section_cluster.py [hunk 25]
```diff
@@ -711,6 +797,7 @@ def test_cluster_section_from_file(mocker, config_parser_dict, expected_dict_par
         ("tags", "{}", {}, None),
         ("tags", "{'test': 'test'}", {"test": "test"}, None),
         ("tags", "{'test': 'test'", None, "Error parsing JSON parameter"),
+        ("disable_hyperthreading", "true", True, None),
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/tests/pcluster/config/test_section_cluster.py]: done

File: cli/tests/pcluster/config/test_validators.py [hunk 26]
```diff
@@ -472,6 +472,28 @@ def test_fsx_storage_capacity_validator(mocker, section_dict, expected_message):
...
+            {"disable_hyperthreading": True, "extra_json": '{"cluster": {"cfn_scheduler_slots": "cores"}}'},
+            "cfn_scheduler_slots cannot be set in addition to disable_hyperthreading = true",
+        ),
+        (
+            {"disable_hyperthreading": True, "extra_json": '{"cluster": {"cfn_scheduler_slots": 3}}'},
+            "cfn_scheduler_slots cannot be set in addition to disable_hyperthreading = true",
+        ),
+    ],
+)
+def test_disable_hyperthreading_validator(mocker, section_dict, expected_message):
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/tests/pcluster/config/test_validators.py]: done

File: cli/pcluster/config/param_types.py [hunk 27]
```diff
@@ -687,6 +695,56 @@ def to_file(self, config_parser):
...
+        """Initialize param value by parsing the right CFN input."""
+        try:
+            cfn_converter = self.definition.get("cfn_param_mapping", None)
+            if cfn_converter and cfn_params:
+                cores = get_cfn_param(cfn_params, cfn_converter)
+                if cores and cores != "NONE":
+                    cores = int(cores.split(",")[0])
+                    self.value = cores > 0
+        except ValueError or IndexError:
+            error("Unable to parse Cfn Parameter Cores = {0}".format(cfn_params))
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/pcluster/config/param_types.py]: ok

File: cli/pcluster/config/param_types.py [hunk 28]
```diff
@@ -687,6 +695,56 @@ def to_file(self, config_parser):
...
+
+        return self
+
+    def to_cfn(self):
+        """
+        Return a tuple (cores_master,cores_compute) if disable_hyperthreading = true.
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/pcluster/config/param_types.py]: changed

File: cli/pcluster/config/param_types.py [hunk 29]
```diff
@@ -687,6 +695,57 @@ def to_file(self, config_parser):
...
+class DisableHyperThreadingParam(BoolParam):
+    """
+    Class to manage the disable_hyperthreading configuration parameter.
+
+    We need this class in order to convert the boolean disable_hyperthreading = [true/false] into Cores.
+    """
+
+    def from_cfn_params(self, cfn_params):
+        """Initialize param value by parsing the right CFN input."""
+        self.value = self.definition.get("default", None)
```
[the remark this answers is not in the document]
Comment (sean-smith) [reply cli/pcluster/config/param_types.py]: Removed. Thanks!
